README.md

# ReactPHP WebDriver

This is a direct port of [RemoteWebDriver](https://github.com/php-webdriver/php-webdriver/blob/1.8.3/lib/Remote/RemoteWebDriver.php)
logic from the [php-webdriver/webdriver](https://github.com/php-webdriver/php-webdriver) package, which utilizes [ReactPHP](https://github.com/reactphp/reactphp)
event loop and promise API for browser interaction w/o execution flow blocking.

**Selenium WebDriver** is a software that is used to manipulate browsers from the code (primarily, for testing and web scraping).
You can find more here: [https://selenium.dev](https://selenium.dev).

This PHP client sends async HTTP requests to the [Grid](https://www.selenium.dev/documentation/en/grid). It is a central
endpoint for commands, a bridge between your code and browser instances. See
[SeleniumHQ/docker-selenium](https://github.com/SeleniumHQ/docker-selenium) to get your own remote browser (or a cluster).

Enjoy!

## Requirements

- **PHP 7.4** or higher.
- ReactPHP v1 (http **^1**, stream **^1**).
- Symfony conflicts: 5.1 (or newer) environments are preferred; the package uses (and will use) some components from
there, and their code / version constraints may need a review, to include a wider range of supported environments
(otherwise, you need to adjust your platform).

## Installation

With [composer](https://getcomposer.org/download):

```
$ composer require itnelo/reactphp-webdriver:^0.4
```

## How to use

Call a factory method to get your instance (recommended). The minimal configuration is:

```php
use React\EventLoop\Factory as LoopFactory;
use Itnelo\React\WebDriver\WebDriverFactory;

$loop = LoopFactory::create();

$webDriver = WebDriverFactory::create(
    $loop,
    [
        'hub' => [
            'host' => 'selenium-hub',
            'port' => 4444,
        ],
    ]
);
```

You can customize a set of parameters for the underlying [ReactPHP Browser](https://github.com/reactphp/http#browser)
and tune driver options:

```php
use React\EventLoop\Factory as LoopFactory;
use Itnelo\React\WebDriver\WebDriverFactory;

$loop = LoopFactory::create();

$webDriver = WebDriverFactory::create(
    $loop,
    [
        'browser' => [
            'tcp' => [
                'bindto' => '192.169.56.10:0',
            ],
            'tls' => [
                'verify_peer' => false,
                'verify_peer_name' => false,
            ],
        ],
        'hub' => [
            'host' => 'selenium-hub',
            'port' => 4444,
        ],
        'command' => [
            'timeout' => 30,
        ],
    ]
);
```

Manual configuration (if you prefer to configure each component as a separate service, e.g. compiling a DI container
and want to reuse existing service definitions):

```php
use React\EventLoop\Factory as LoopFactory;
use React\Socket\Connector as SocketConnector;
use React\Http\Browser;
use Itnelo\React\WebDriver\Client\W3CClient;
use Itnelo\React\WebDriver\Timeout\Interceptor as TimeoutInterceptor;
use Itnelo\React\WebDriver\SeleniumHubDriver;

$loop = LoopFactory::create();

$socketConnector = new SocketConnector(
    $loop,
    [
        'tcp' => [
            'bindto' => '192.169.56.10:0',
        ],
        'tls' => [
            'verify_peer' => false,
            'verify_peer_name' => false,
        ],
    ]
);
$browser = new Browser($loop, $socketConnector);
$browser = $browser->withRejectErrorResponse(false);

$hubClient = new W3CClient(
    $browser,
    [
        'server' => [
            'host' => 'selenium-hub',
            'port' => 4444,
        ],
    ]
);

$timeoutInterceptor = new TimeoutInterceptor($loop, 30);

$webDriver = new SeleniumHubDriver(
    $loop,
    $hubClient,
    $timeoutInterceptor
);
```

See a self-documented [WebDriverInterface.php](src/WebDriverInterface.php) (and [ClientInterface.php](src/ClientInterface.php))
for the API details. Not all methods and arguments are ported (only the most necessary), so feel free to open
an issue / make a pull request if you want more.

## See also

- [php-webdriver/webdriver](https://github.com/php-webdriver/php-webdriver) — the original, "blocking" implementation;
to get information how to use some advanced methods. For example, [WebDriverKeys](https://github.com/php-webdriver/php-webdriver/blob/main/lib/WebDriverKeys.php#L10)
helper describes Unicode strings for sending special inputs to page elements (e.g. `Ctrl`, `Alt` and other keys).

## Changelog

All notable changes to this project will be documented in [CHANGELOG.md](CHANGELOG.md).

CHANGELOG.md

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- No description yet.

## [0.4.0] - 2021-02-25

### Added

- `ClientInterface::getSessionIdentifiers()` implementation to get a list with WebDriver sessions from the active pool
(as an opportunity to resume existing sessions).

### Fixed

- Missed type checks after WebDriver response deserialization for some methods (e.g. to prevent misreading an
error array as a valid command output).

## [0.3.0] - 2021-02-11

### Added

- `ClientInterface::removeSession()` implementation to properly close a WebDriver session.
- Configurable `checkInterval` for the `WebDriverInterface::waitUntil()` method, to define a custom frequency for the
condition checks.
- README: minimal configuration snippet.

### Fixed

- `command.timeout` option now accepts both float and int values (will be normalized to float by the driver factory).
- `onCommandConfirmation` hook is now propagates driver error message from the response alongside the predefined error
label.

## [0.2.1] - 2020-12-22

### Fixed

- `WebDriverInterface::waitUntil()`: a resulting value from the promise, that has been generated by the
"condition-to-satisfy" callback, is now properly forwarded to the `onFulfilled` handler, for the `waitUntil` method
itself.

## [0.2.0] - 2020-12-18

### Added

- `ClientInterface::getTabIdentifiers()` method implementation (reading a list of tabs, which are opened in the
browser).
- Implementations for the `ClientInterface::getActiveTabIdentifier()` and `ClientInterface::setActiveTab()` commands,
to determine which of the tabs is currently "focused" and to open (focus) a specific one.
- `ClientInterface::openUri()` and `ClientInterface::getCurrentUri()` implementations, to open a web page
in the current browser tab and read an URI from the address bar.
- `ClientInterface::getSource()` for acquiring source code of the web resource that is opened in the active (focused)
browser tab (view-source action).
- `ClientInterface::getElementIdentifier()` to get an internal WebDriver handle of the specific element on the page.
- `ClientInterface::getActiveElementIdentifier()` to get an internal handle of the currently focused element.
- `ClientInterface::getElementVisibility()` to check if the target is visible on the page.
- `ClientInterface::mouseMove()` and `ClientInterface::mouseLeftClick()` to perform the most common mouse actions in
the browser window.
- `ClientInterface::keypressElement()` to send keyboard keys to the element on the page.
- `ClientInterface::getScreenshot()` and a `WebDriverInterface::saveScreenshot()` shortcut implementation, to capture
a screen state of the remote browser instance.
- `WebDriverInterface::wait()` and `WebDriverInterface::waitUntil()` to make delays and conditional checks
(waiting for element visibility, etc.).
- README: a "Requirements" block, description improvements.

### Changed

- `Timeout\Interceptor` is now raising a single exception with both method-specific message and a timeout context.

### Fixed

- Interface fixes and a minor redesign for some signatures.

This version brings an async variant for most methods (no tests yet), which were defined by the
`WebDriverInterface` and `ClientInterface`, to communicate with a remote Selenium WebDriver service in non-blocking
mode, except: `getSessionIdentifiers`, `removeSession` and `clickElement` commands (the last one can be supplanted
by the `mouseLeftClick`, in combination with the `mouseMove` action).

At this moment, `ClientInterface::createSession()` is able to send only a hardcoded set of capabilities (settings
for the executable file and profile preferences) to launch a browser instance (will be patched in the later versions):

```
{
    "browserName": "chrome",
    "goog:chromeOptions": {
        "args": [
            "--user-data-dir=\u002fopt\u002fgoogle\u002fchrome\u002fprofiles"
        ],
        "prefs": {
            "intl.accept_languages": "RU-ru,ru,en-US,en"
        }
    }
}
``` 

## [0.1.0] (core design) - 2020-12-02

### Added

- `WebDriverInterface` and low-level `ClientInterface` to communicate with
[Selenium Grid](https://www.selenium.dev/documentation/en/grid) server asynchronously, using the centralized
[event loop](https://github.com/reactphp/event-loop) and [promise API](https://github.com/reactphp/promise).
- `SeleniumHubDriver` and `Client\W3CClient` stubs for W3C compliant webdriver implementation.
- `WebDriverFactory` as a shortcut for driver instantiation.
- `Timeout\Interceptor` to prevent unresolved (hanging) driver promises, whenever it fails
(using [reactphp/promise-timer](https://github.com/reactphp/promise-timer)).
- `ClientInterface::createSession()` method implementation (opening a Selenium hub session to interact with remote
browser instance).

This early development version doesn't yet contain full implementation for the introduced `WebDriverInterface`, only
core design solutions and library interfaces are defined.

[Unreleased]: https://github.com/itnelo/reactphp-webdriver/compare/0.4.0...0.x
[0.4.0]: https://github.com/itnelo/reactphp-webdriver/compare/0.3.0..0.4.0
[0.3.0]: https://github.com/itnelo/reactphp-webdriver/compare/0.2.1..0.3.0
[0.2.1]: https://github.com/itnelo/reactphp-webdriver/compare/0.2.0..0.2.1
[0.2.0]: https://github.com/itnelo/reactphp-webdriver/compare/0.1.0..0.2.0
[0.1.0]: https://github.com/itnelo/reactphp-webdriver/releases/tag/0.1.0

src/ClientInterface.php
<?php

/*
 * This file is part of the ReactPHP WebDriver <https://github.com/itnelo/reactphp-webdriver>.
 *
 * (c) 2020 Pavel Petrov <itnelo@gmail.com>.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @license https://opensource.org/licenses/mit MIT
 */

declare(strict_types=1);

namespace Itnelo\React\WebDriver;

use React\Promise\PromiseInterface;

/**
 * The async client for Selenium Grid server (hub)
 *
 * @see https://github.com/SeleniumHQ/docker-selenium
 */
interface ClientInterface
{
    /**
     * Returns a promise that resolves to collection of session identifiers (ex. '0dd27f03d50e795f8499d1c99f7c01dd'),
     * which are available for command execution.
     *
     * Each session represent an opened browser, with some user profile and tabs (window handles).
     *
     * Resulting collection represents a Traversable<string> or string[].
     *
     * @return PromiseInterface<iterable>
     */
    public function getSessionIdentifiers(): PromiseInterface;

    /**
     * Returns a promise that resolves to a string, representing identifier of session that has been started by the
     * remote WebDriver service.
     *
     * Usage example:
     *
     * ```
     * $sessionIdentifierPromise = $webDriver->createSession();
     *
     * $sessionIdentifierPromise->then(
     *     function (string $sessionIdentifier) {
     *         // do some work
     *     },
     *     function (Throwable $rejectionReason) {
     *         // handle exception
     *     }
     * );
     * ```
     *
     * todo: argument to send desired capabilities, i.e. browser arguments and preferences (omitted)
     * todo: must also return a collection of confirmed browser capabilities (omitted)
     *
     * @return PromiseInterface<string>
     */
    public function createSession(): PromiseInterface;

    /**
     * Returns a promise that will be resolved when the remote WebDriver service confirms closed session.
     *
     * Such session can't be used (or resumed) after this promise is successfully resolved.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<null>
     */
    public function removeSession(string $sessionIdentifier): PromiseInterface;

    /**
     * Returns a promise that resolves to collection of tab identifiers (i.e. "window handles") that currently are
     * opened in the browser.
     *
     * Resulting collection represents a string[].
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<array>
     */
    public function getTabIdentifiers(string $sessionIdentifier): PromiseInterface;

    /**
     * Returns a promise that resolves to a string, representing an identifier of browser tab that currently is active
     * (focused) and available for interaction (ex. 'CDwindow-A4F079B43D7E1D267F6D5F11FE8CC374')
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<string>
     */
    public function getActiveTabIdentifier(string $sessionIdentifier): PromiseInterface;

    /**
     * Returns a promise that will be resolved when the remote WebDriver service confirms browser tab switching using a
     * given identifier (window handle).
     *
     * Note: you can set to active state only existing tabs (see {@link getTabIdentifiers()}).
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param string $tabIdentifier     Window handle for "tab" switching operation
     *
     * @return PromiseInterface<null>
     */
    public function setActiveTab(string $sessionIdentifier, string $tabIdentifier): PromiseInterface;

    /**
     * Returns a promise that will be resolved when the remote WebDriver service confirms URI navigation within
     * currently active (focused) tab ({@link getActiveTabIdentifier()})
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param string $uri               Website URL or any other resource identifier, to open in the active (focused)
     *                                  browser tab
     *
     * @return PromiseInterface<null>
     */
    public function openUri(string $sessionIdentifier, string $uri): PromiseInterface;

    /**
     * Returns a promise that resolves to a string, representing an identifier of web resource, which is opened in the
     * currently active browser tab (ex. 'https://github.com/itnelo')
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<string>
     */
    public function getCurrentUri(string $sessionIdentifier): PromiseInterface;

    /**
     * Returns a promise that resolves to a string, representing source code of web resource, which is opened in the
     * currently active browser tab (view-source action)
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<string>
     */
    public function getSource(string $sessionIdentifier): PromiseInterface;

    /**
     * Returns a promise that resolves to an array, representing identifier (an internal handle) of the first element
     * on page, found by the given XPath query (currently active browser tab is used).
     *
     * Resulting value example:
     * ```
     * [
     *     'element-6066-11e4-a52e-4f735466cecf' => '1aab1034-814b-4a0b-95cb-94c7c58bbf7a'
     * ]
     * ```
     *
     * Use this handle to perform element-specific actions (e.g. {@link clickElement()} and {@link keypressElement()}).
     *
     * Note: this method behavior equals to "findElement" from the original (blocking) implementation, so if element is
     * not present in the DOM (or a given XPath query is invalid), promise will be rejected with the corresponding
     * RuntimeException as a reason.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param string $xpathQuery        XPath query to find an element in the DOM tree (e.g. '//input[@type="text"]')
     *
     * @return PromiseInterface<array>
     */
    public function getElementIdentifier(string $sessionIdentifier, string $xpathQuery): PromiseInterface;

    /**
     * Returns a promise that resolves to an array, representing identifier (an internal handle) of the currently
     * active (focused) element on page.
     *
     * Resulting value has the same representation as {@link getElementIdentifier()} resval.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<array>
     */
    public function getActiveElementIdentifier(string $sessionIdentifier): PromiseInterface;

    /**
     * Returns a promise that resolves to a boolean, representing element visibility status on the currently active
     * browser tab.
     *
     * Similar to isDisplayed() method from the original (blocking) HttpCommandExecutor.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param array  $elementIdentifier An internal WebDriver handle that refers to the element on the page
     *
     * @return PromiseInterface<bool>
     */
    public function getElementVisibility(string $sessionIdentifier, array $elementIdentifier): PromiseInterface;

    /**
     * Returns a promise that will be resolved when the remote WebDriver service confirms element click operation
     * against active browser tab.
     *
     * Note: if target element is a link, it may trigger an implicit tab(s) creation. In such case you need to take
     * into consideration potentially new list of window handles for the current session (see {@link
     * getTabIdentifiers()}). The remote webdriver service WILL NOT automatically perform tab switch operation if
     * a command implicitly triggers new tab.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param array  $elementIdentifier An internal WebDriver handle that refers to the element on the page
     *
     * @return PromiseInterface<null>
     */
    public function clickElement(string $sessionIdentifier, array $elementIdentifier): PromiseInterface;

    /**
     * Returns a promise that will be resolved when the remote WebDriver service finishes key sequence apply operation
     * to the specified page element on the active browser tab.
     *
     * Note: you can't send a key sequence without targeting a concrete element on the page. The original (blocking)
     * implementation sends descriptor of currently active element as an implicit target for this case.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param array  $elementIdentifier An internal WebDriver handle that refers to the element on the page
     * @param string $keySequence       Text to type or a sequence of special symbols (e.g. "\xEE\x80\x89" for CTRL)
     *
     * @return PromiseInterface<null>
     *
     * @see https://github.com/php-webdriver/php-webdriver/blob/1.8.3/lib/WebDriverKeys.php#L10
     */
    public function keypressElement(
        string $sessionIdentifier,
        array $elementIdentifier,
        string $keySequence
    ): PromiseInterface;

    /**
     * Returns a promise that will be resolved when the remote WebDriver service finishes mouse moving operation
     * against a specified point on the page.
     *
     * Note 1: if you don't specify {startingPoint}, an active (focused) element will be used implicitly (see
     * {@link getActiveElement()}. If no active element (ex. page just loaded) - {0, 0} will be used. An internal
     * pointer of type "mouse" will PRESERVE its coordinates between page transition on the same tab (see a problem
     * description below).
     *
     * Note 2: if {offsetX} or {offsetY} is too high, you may get an "out of bounds" error, which means target
     * point is not in the browser window. The workaround is to always set a proper starting point (e.g. some element
     * at the top-left corner of the page - if you planning to make some detection-breaking moves) and knowing your
     * boundaries.
     *
     * @param string     $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param int        $offsetX           A vertical offset that defines mouse move direction (from top to bottom)
     * @param int        $offsetY           A horizontal offset (from left to right)
     * @param int        $moveDuration      How fast WebDriver will move an internal pointer (original, blocking
     *                                      implementation has a 100 ms time frame dy default)
     * @param array|null $startingPoint     An element identifier (internal WebDriver handle) as a starting point, from
     *                                      which mouse will be moved by the given {offsetX} and {offsetY}
     *
     * @return PromiseInterface<null>
     */
    public function mouseMove(
        string $sessionIdentifier,
        int $offsetX,
        int $offsetY,
        int $moveDuration = 100,
        array $startingPoint = null
    ): PromiseInterface;

    /**
     * Returns a promise that will be resolved when the remote WebDriver service confirms both "pointerDown" and
     * "pointerUp" actions for left button of the virtual mouse (button 0), at the current position of internal pointer
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<null>
     */
    public function mouseLeftClick(string $sessionIdentifier): PromiseInterface;

    /**
     * Returns a promise that resolves to a string, raw content of the image, generated by the remote WebDriver service
     * and representing visual state of currently active (focused) browser tab.
     *
     * The resulting image is not automatically saved anywhere and user-side must perform persisting logic (PNG
     * format), if needed.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     *
     * @return PromiseInterface<string>
     */
    public function getScreenshot(string $sessionIdentifier): PromiseInterface;
}


src/WebDriverInterface.php
<?php

/*
 * This file is part of the ReactPHP WebDriver <https://github.com/itnelo/reactphp-webdriver>.
 *
 * (c) 2020-2021 Pavel Petrov <itnelo@gmail.com>.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @license https://opensource.org/licenses/mit MIT
 */

declare(strict_types=1);

namespace Itnelo\React\WebDriver;

use React\Promise\PromiseInterface;

/**
 * Manipulates a remote browser instance asynchronously, using Selenium Grid (hub) API.
 *
 * As an extension to the base client, it can perform both action requests and some client-side (offline) tasks,
 * between these actions, such as conditional waits.
 */
interface WebDriverInterface extends ClientInterface
{
    /**
     * Returns a promise that will be resolved when the driver completes idling for the specified amount of time.
     *
     * Usage example:
     *
     * ```
     * $navigationPromise = $webDriver->openUri('https://github.com/itnelo');
     *
     * $elementIdentifierPromise = $navigationPromise->then(
     *     function () use ($webDriver) {
     *         // try-catch
     *         $timeHasComePromise = $webDriver->wait(5.0);
     *
     *         return $timeHasComePromise->then(
     *             function () use ($webDriver) {
     *                 return $webDriver->getElementIdentifier('sessionIdentifier', 'xpathQuery');
     *             }
     *         );
     *     }
     *     // handle rejection reason (e.g. a connection timeout due to unexpected rate limiting)
     * );
     *
     * $elementClickPromise = $elementIdentifierPromise->then(
     *     function (string $elementIdentifier) use ($webDriver) {
     *         // try-catch
     *         return $webDriver->clickElement('sessionIdentifier', $elementIdentifier);
     *     }
     *     // handle rejection reason (e.g. invalid xpath or element not found error)
     * );
     * ```
     *
     * Note: each wait call is a separate action and starts its own timer; that timer will not wait for any other
     * timers to fire (i.e. this method is not suited for "shooting" an array of concurrent requests with delays, to
     * bypass rate limits, but to wait page loading and javascript code). Basically, it is just a syntactic sugar for
     * promise timer boilerplate, so you should use the event reactor directly, to limit driver calls for some
     * sensitive operations.
     *
     * @param float $time Time in seconds to wait (e.g. 0.351; max precision can be 3)
     *
     * @return PromiseInterface<null>
     */
    public function wait(float $time = 30.0): PromiseInterface;

    /**
     * Returns a promise that will be resolved, when a given condition is met within specified amount of time, and
     * rejected, otherwise.
     *
     * A condition callback must return an instance of PromiseInterface. Whenever that promise becomes rejected, driver
     * will try to get a new promise from the callback, until it reaches a given timeout for retry attempts.
     *
     * Note 1: waitUntil itself doesn't apply any timeouts for a single result promise (check iteration), the user-side
     * MUST control any kind of promise timeouts for the condition callback, which is supplied to the method.
     *
     * Note 2: the resulting value for the waitUntil promise will be a value, which is forwarded by the condition
     * callback (see an example below).
     *
     * Usage example:
     *
     * ```
     * $becomeVisiblePromise = $webDriver->waitUntil(
     *     function () use ($webDriver) {
     *         $visibilityStatePromise = $webDriver->getElementVisibility(...);
     *
     *         return $visibilityStatePromise->then(
     *             function (bool $isVisible) {
     *                 if (!$isVisible) {
     *                     throw new RuntimeException("Not visible yet! Let's retry!");
     *                 }
     *
     *                 // this value will be forwarded as a "result value" for the waitUntil promise.
     *                 return 'it is visible now!';
     *             }
     *         );
     *     },
     *     15.5
     * );
     *
     * $becomeVisiblePromise->then(
     *     function (string $forwardedValue) use ($webDriver) {
     *         // try-catch
     *         $webDriver->clickElement(...);    // sending a click command only if we are sure the target is visible.
     *
     *         // $forwardedValue === 'it is visible now!'
     *     }
     *     // handle case when the element is not visible on the page
     * );
     * ```
     *
     * @param callable $conditionMetCallback A condition to be met, as a callback
     * @param float    $time                 Time (in seconds) to wait for successfully resolved promise from the
     *                                       condition callback (minimum: 0.5)
     * @param float    $checkInterval        The interval for condition checks, in seconds (minimum: 0.1)
     *
     * @return PromiseInterface<mixed>
     */
    public function waitUntil(
        callable $conditionMetCallback,
        float $time = 30.0,
        float $checkInterval = 0.5
    ): PromiseInterface;

    /**
     * Returns a promise that will be resolved if a screenshot is successfully received and saved using the specified
     * {filePath}, rejection reason with error message will be provided otherwise.
     *
     * @param string $sessionIdentifier Session identifier for Selenium Grid server (hub)
     * @param string $filePath          Path, where a screenshot image must be saved
     *
     * @return PromiseInterface<null>
     */
    public function saveScreenshot(string $sessionIdentifier, string $filePath): PromiseInterface;
}


src/WebDriverFactory.php
<?php

/*
 * This file is part of the ReactPHP WebDriver <https://github.com/itnelo/reactphp-webdriver>.
 *
 * (c) 2020 Pavel Petrov <itnelo@gmail.com>.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @license https://opensource.org/licenses/mit MIT
 */

declare(strict_types=1);

namespace Itnelo\React\WebDriver;

use Itnelo\React\WebDriver\Client\W3CClient;
use Itnelo\React\WebDriver\Timeout\Interceptor as TimeoutInterceptor;
use React\EventLoop\LoopInterface;
use React\Http\Browser;
use React\Socket\Connector as SocketConnector;
use Symfony\Component\OptionsResolver\Exception\ExceptionInterface as ConfigurationExceptionInterface;
use Symfony\Component\OptionsResolver\Options;
use Symfony\Component\OptionsResolver\OptionsResolver;

/**
 * Builds web driver instances by the given configuration settings
 */
final class WebDriverFactory
{
    /**
     * Creates and returns a new web driver instance.
     *
     * Usage example:
     *
     * ```
     * $loop = \React\EventLoop\Factory::create();
     *
     * $webDriver = \Itnelo\React\WebDriver\WebDriverFactory::create(
     *     $loop,
     *     [
     *         'browser' => [
     *             'tcp' => [
     *                 'bindto' => '192.169.56.10:0',
     *             ],
     *             'tls' => [
     *                 'verify_peer' => false,
     *                 'verify_peer_name' => false,
     *             ],
     *         ],
     *         'hub' => [
     *             'host' => 'selenium-hub',
     *             'port' => 4444,
     *         ],
     *         'command' => [
     *             'timeout' => 30,
     *         ],
     *     ]
     * );
     * ```
     *
     * For all available "browser" options see \React\Socket\Connector class (will be instantiated for the underlying
     * browser) and socket context options: https://php.net/manual/en/context.socket.php.
     *
     * The "command.timeout" option here doesn't correlate with ReactPHP Browser's timeouts and will just cancel a
     * pending promise after the specified time (in seconds). Furthermore, the hub client can reject promise with a
     * runtime exception if an underlying browser has decided to stop waiting for the response by its own settings.
     *
     * @param LoopInterface $loop    The event loop reference to create underlying components
     * @param array         $options An array of configuration settings for the driver
     *
     * @return WebDriverInterface
     *
     * @throws ConfigurationExceptionInterface Whenever an error has been occurred during driver instantiation
     */
    public static function create(LoopInterface $loop, array $options): WebDriverInterface
    {
        $optionsResolver = new OptionsResolver();

        $optionsResolver
            ->define('browser')
            ->info('Options to customize a socket connector, which will be used by the internal http client')
            ->default(
                function (OptionsResolver $browserOptionsResolver) {
                    $browserOptionsResolver->setDefined(['tcp', 'tls', 'unix', 'dns', 'timeout', 'happy_eyeballs']);
                }
            )
        ;

        $optionsResolver
            ->define('hub')
            ->info('Options to create a hub client that will send commands to the Selenium Grid endpoint')
            ->default(
                function (OptionsResolver $hubOptionsResolver) {
                    $hubOptionsResolver->setDefined(['host', 'port']);
                }
            )
        ;

        $optionsResolver
            ->define('command')
            ->info('Options to control behavior of the commands, which will be executed on the remote server')
            ->default(
                function (OptionsResolver $commandOptionsResolver) {
                    $commandOptionsResolver
                        ->define('timeout')
                        ->info(
                            'Maximum time to wait (in seconds) for command execution '
                            . '(do not correlate with HTTP timeouts)'
                        )
                        ->allowedTypes('int', 'float')
                        ->default(30.0)
                        ->normalize(
                            function (Options $options, $commandTimeout) {
                                $commandTimeoutNormalized = (float) $commandTimeout;
                                $commandTimeoutNormalized = max(0.1, $commandTimeoutNormalized);

                                return $commandTimeoutNormalized;
                            }
                        )
                    ;
                }
            )
        ;

        $optionsResolved = $optionsResolver->resolve($options);

        $socketConnector = new SocketConnector($loop, $optionsResolved['browser']);
        $httpClient      = new Browser($loop, $socketConnector);

        // Selenium hub sends some valid responses with 5xx status codes, so we need to disable eager promise rejection
        // to properly parse error messages and other details from the body.
        $httpClient = $httpClient->withRejectErrorResponse(false);

        $hubClient          = new W3CClient($httpClient, ['server' => $optionsResolved['hub']]);
        $timeoutInterceptor = new TimeoutInterceptor($loop, $optionsResolved['command']['timeout']);

        $webDriverOptions = [];
        $webDriver        = new SeleniumHubDriver($loop, $hubClient, $timeoutInterceptor, $webDriverOptions);

        return $webDriver;
    }
}
